core/trino-main/src/main/java/io/trino/operator/DeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,348 @@
+    {
+        if (failure != null) {
+            return;
+        }
+
+        if (inputFinished) {
+            return;
+        }
+
+        if (!noMoreTasks) {
+            return;
+        }
+
+        if (allTasks.isEmpty()) {
+            inputFinished = true;
+            unblockIfNecessary(blocked);
+            return;
+        }
+
+        switch (retryPolicy) {
+            case TASK:
+                // TODO implement deduplication for task level retires
+                throw new UnsupportedOperationException("task level retry policy is unsupported");
+            case QUERY: {
+                int latestAttemptId = allTasks.stream()
+                        .mapToInt(TaskId::getAttemptId)
+                        .max()
+                        .orElseThrow(() -> new VerifyException("allTasks is not expected to be empty"));
+
+                Set<TaskId> latestAttemptTasks = allTasks.stream()
+                        .filter(taskId -> taskId.getAttemptId() == latestAttemptId)
+                        .collect(toImmutableSet());
+
+                if (finishedTasks.containsAll(latestAttemptTasks)) {
+                    removePagesForPreviousAttempts(latestAttemptId);
+                    inputFinished = true;
+                    unblockIfNecessary(blocked);
+                }
+
+                List<Throwable> failures = failedTasks.entrySet().stream()
+                        .filter(entry -> entry.getKey().getAttemptId() == latestAttemptId)
+                        .map(Map.Entry::getValue)
+                        .collect(toImmutableList());
+
+                if (!failures.isEmpty()) {
+                    Throwable failure = null;
+                    for (Throwable taskFailure : failures) {
+                        if (failure == null) {
+                            failure = taskFailure;
+                        }
+                        else if (failure != taskFailure) {
+                            failure.addSuppressed(taskFailure);
+                        }
+                    }
+                    pageBuffer.clear();
+                    bufferRetainedSizeInBytes = 0;
+                    this.failure = failure;
+                    unblockIfNecessary(blocked);
+                }
+                break;
+            }
+            default:
+                throw new VerifyException("unexpected retry policy");
+        }
+    }
+
+    private synchronized void removePagesForPreviousAttempts(int currentAttemptId)
+    {
+        // wipe previous attempt pages
+        long pagesRetainedSizeInBytes = 0;
+        Iterator<Map.Entry<TaskId, SerializedPage>> iterator = pageBuffer.entries().iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<TaskId, SerializedPage> entry = iterator.next();
+            if (entry.getKey().getAttemptId() < currentAttemptId) {
+                pagesRetainedSizeInBytes += entry.getValue().getRetainedSizeInBytes();
+                iterator.remove();
```

Comment:
same here; it may be costly. You may consider:
 * integrate over keys
 * compute total size of pages to be dropped for given key
 * remove all entries for given key at once using `pageBuffer.removeAll`

Reply:
Switched to a LinkedList based implementation for now.